types/decimal.go

```diff
@@ -0,0 +1,356 @@
+
+// multiplication
+func (d Dec) Mul(d2 Dec) Dec {
+	mul := new(big.Int).Mul(d.Int, d2.Int)
+	chopped := BankerRoundChop(mul, Precision)
+
+	if chopped.BitLen() > 255 {
+		panic("Int overflow")
+	}
+	return Dec{chopped}
+}
+
+// quotient
+func (d Dec) Quo(d2 Dec) Dec {
+
+	// multiply precision twice
+	mul := new(big.Int).Mul(d.Int, precisionReuse)
+	mul.Mul(mul, precisionReuse)
+
+	quo := new(big.Int).Quo(mul, d2.Int)
+	chopped := BankerRoundChop(quo, Precision)
+	return Dec{chopped}
+}
+
+func (d Dec) String() string {
+	str := d.ToLeftPaddedWithDecimals(Precision)
+	placement := len(str) - Precision
+	if placement < 0 {
+		panic("too few decimal digits")
+	}
+	return str[:placement] + "." + str[placement:]
+}
+
+// TODO panic if negative or if totalDigits < len(initStr)???
+// evaluate as an integer and return left padded string
+func (d Dec) ToLeftPaddedWithDecimals(totalDigits int8) string {
+	intStr := d.Int.String()
+	fcode := `%0` + strconv.Itoa(int(totalDigits)) + `s`
+	return fmt.Sprintf(fcode, intStr)
+}
+
+// TODO panic if negative or if totalDigits < len(initStr)???
+// evaluate as an integer and return left padded string
+func (d Dec) ToLeftPadded(totalDigits int8) string {
+	chopped := BankerRoundChop(d.Int, Precision)
+	intStr := chopped.String()
+	fcode := `%0` + strconv.Itoa(int(totalDigits)) + `s`
+	return fmt.Sprintf(fcode, intStr)
+}
+
+//     ____
+//  __|    |__   "chop 'em
+//       ` \     round!"
+// ___||  ~  _     -bankers
+// |         |      __
+// |       | |   __|__|__
+// |_____:  /   | $$$    |
+//              |________|
+
+// nolint - go-cyclo
+// chop of n digits, and banker round the digits being chopped off
+// Examples:
+//   BankerRoundChop(1005, 1) = 100
+//   BankerRoundChop(1015, 1) = 102
+//   BankerRoundChop(1500, 3) = 2
+func BankerRoundChop(d *big.Int, n int64) (chopped *big.Int) {
+
+	// remove the negative and add it back when returning
+	if d.Sign() == -1 {
+		d = new(big.Int).Neg(d)
+		defer func() {
+			chopped = new(big.Int).Neg(chopped)
+		}()
+	}
+
+	// get the trucated quotient and remainder
+	quo, rem := big.NewInt(0), big.NewInt(0)
+	quo, rem = quo.QuoRem(d, precisionReuse, rem)
+
+	if rem.Sign() == 0 { // remainder is zero
+		return quo
+	}
+
+	lenWhole := len(d.String())
+	if quo.Sign() == 0 { // only the decimal places (ex. 0.1234)
+		lenWhole++
+	}
+	lenQuo := len(quo.String())
+	lenRem := len(rem.String())
+	leadingZeros := lenWhole - (lenQuo + lenRem) // leading zeros removed from the remainder
+
+	zerosToAdd := int64(lenRem - 1 + leadingZeros)
+	multiplier := new(big.Int).Exp(big.NewInt(10), big.NewInt(zerosToAdd), nil)
+	fiveLine := new(big.Int).Mul(big.NewInt(5), multiplier)
+
+	switch rem.Cmp(fiveLine) {
+	case -1:
+		chopped = quo
+		return
+	case 1:
+		chopped = new(big.Int).Add(quo, big.NewInt(1))
+		return
+	default: // bankers rounding must take place
+		str := quo.String()
+		finalDig, err := strconv.Atoi(string(str[len(str)-1]))
```

Comment:
Why do we need to convert to a string and then to an integer - can't we just check the byte value directly?

Reply:
addressed by val